ludwig/utils/strings_utils.py

```diff
@@ -243,6 +243,84 @@ class Vocabulary:
     unknown_symbol: str
     """Actual unknown symbol."""
 
+    prompt_template_num_tokens: int = 0
+    """The number of tokens in the prompt template.
+
+    If -1, then there is no prompt template.
+    """
+
+
+def _get_vocabulary(
+    tokenizer_type: str,
+    tokenizer,
+    vocab_file: str,
+    unknown_symbol: str,
+    add_special_symbols: bool,
+    padding_symbol: str,
+    unit_counts: Counter,
+    num_most_frequent: int,
+) -> Optional[List[str]]:
+    """Returns the vocabulary from the tokenizer_type, tokenizer, or vocab_file.
+
+    If there's no vocab_file or if the tokenizer has no set vocabulary (e.g. space_punct), then the vocabulary is
+    determined from the tokenized data.
+    """
+    # Pre-trained huggingface tokenizer. Use the pre-existing vocabulary and special symbols.
+    if tokenizer_type == "hf_tokenizer":
+        try:
+            vocab = tokenizer.get_vocab()
+            vocab = list(vocab.keys())
+        except NotImplementedError:
+            logger.warning(
+                "HuggingFace tokenizer does not have a get_vocab() method. "
+                + "Using tokenizer.tokenizer.vocab_size and tokenizer.tokenizer._convert_id_to_token "
+                + "to build the vocabulary."
+            )
+            vocab = []
+            for idx in range(tokenizer.tokenizer.vocab_size):
+                vocab.append(tokenizer.tokenizer._convert_id_to_token(idx))
+            vocab += tokenizer.tokenizer.added_tokens_encoder.keys()
+
+        pad_token = tokenizer.get_pad_token()
+        unk_token = tokenizer.get_unk_token()
+
+        if unk_token is None:
+            logger.warning(
+                "No unknown token found in HuggingFace tokenizer. Adding one. "
+                + "NOTE: This will change the vocabulary size and may affect model "
+                + "performance, particularly if the model weights are frozen."
+            )
+            vocab = [unknown_symbol] + vocab
+        else:
+            unknown_symbol = unk_token
+
+        if pad_token is None and add_special_symbols:
+            logger.warning(
+                "No padding token found in HuggingFace tokenizer. Adding one. "
+                + "NOTE: This will change the vocabulary size and may affect model "
+                + "performance, particularly if the model weights are frozen."
+            )
+            vocab = [padding_symbol] + vocab
+        else:
+            padding_symbol = pad_token
```

Comment:
Just a general thought - adding these tokens to the vocab is likely going to cause some problems at training time because the embedding layer in pretrained text encoder assumes a fixed vocab size (assuming training from pretrained instead of pertaining from scratch) as one of the arguments in the Embedding layer initialization, so hopefully in our HFTokenizer class implementation, we always set the pad and unk token so that 

```python
pad_token = tokenizer.get_pad_token()
unk_token = tokenizer.get_unk_token()
```

are never none.

Reply:
I just took a look at our implementation. I want to say that unk_token will likely always be set, and our HFTokenizer class does a best effort setting of the pad token with a resolve to set the `pad_token` to token ID 0. I'd say that I've also missed this and am just realizing it, but we probably also want to set tokenizer.pad_token to the corresponding vocab token mapping to token_id 0 after [this](https://github.com/ludwig-ai/ludwig/blob/305fdd2d921c39baba104bdc4e769730ee4b6fad/ludwig/utils/tokenizers.py#L882). 

Otherwise since `tokenizer.get_pad_token()` just returns `self.tokenizer.pad_token`, it may actually return an unknown here and try and insert a different pad token into the vocab.

Reply:
I know this isn't net new code and that you just refactored the create_vocabulary() function into this new private helper method, but just wanted to flag it anyway! We can come back to it in a future PR, of it's easy to do, could be worth just adding a line to also set the pad_token to the corresponding pad_token_id mapping from the vocab.

Reply:
Discussed offline. Pushed up this commit to ensure that pad token and pad token id is always available for HF tokenizers, and simplify the special symbol special handling in vocab generation: https://github.com/ludwig-ai/ludwig/pull/3719/commits/e9b4666d701d6835c0e100888287e22b515a0540